tephra-core/src/main/java/co/cask/tephra/visibility/VisibilityBarrier.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.tephra.visibility;
+
+import co.cask.tephra.TransactionAware;
+
+/**
+ * VisibilityBarrier is used to ensure that after a given point in time, all readers see an updated change
+ * that got committed.
+ * <p>
+ *
+ * Typically a reader will never conflict with a writer, since a reader only sees committed changes when its
+ * transaction started. However to ensure that after a given point all readers are aware of a change,
+ * we have to introduce a conflict between a reader and a writer that act on the same data concurrently.
+ *<p>
+ *
+ * This is done by the reader indicating that it is interested in changes to a piece of data by using a read barrier
+ * in its transaction. If there are no changes to the data when reader completes the transaction
+ * with the read barrier, the commit succeeds.
+ * <p>
+ * After a writer is done updating the data that a reader is interested in, it tries to commit the write barrier in a
+ * separate transaction.
+ * If the writer is able to successfully commit the transaction with write barrier, it means that
+ * any in-progress readers that have not seen the change will not be allowed to commit anymore.
+ * In case an in-progress reader commits before the write barrier is committed, then
+ * the writer will get a conflict on commit, and the writer has to retry the write barrier with a new transaction
+ * until it successfully commits.
+ * <p>
+ * Any in-progress readers after the write barrier is committed will get a conflict when they try to commit. This will
+ * force the reader to start a new transaction, and this ensures that the changes made by writer are visible
+ * to the reader.
+ * <p>
+ * Hence a successful commit of a write barrier ensures that any read barrier that successfully commits after
+ * this point onwards would have seen the change.
+ *
+ * <p>
+ * Sample reader code:
+ * <pre>
+ *   <code>
+ * TransactionAware readBarrier = VisibilityBarrier.reader(barrierId);
```

Comment:
A concrete use case that explains how one can construct the barrierId might make it a lot easier to understand. Otherwise LGTM :+1: